lib/cli/watch-run.js

```diff
@@ -137,40 +140,228 @@ exports.watchRun = (mocha, {watchFiles, watchIgnore}, fileCollectParams) => {
+    // since it's possible for the last path segment to be a directory
+    if (!absPath.endsWith(globEnd)) {
+      res.match.globs.add(path.resolve(absPath, '**', '*'));
+    }
+  }
+
+  debug("returning path filter: %o", res);
+  return res;
+}
+
+/**
+ * Checks if the provided path matches with the path pattern.
+ * @param {string} filePath The path to match.
+ * @param {PathPattern} pattern The path pattern for matching.
+ * @param {boolean} [matchParent] Treats the provided path as a match if it's a valid parent directory from the list of paths.
+ * @returns {boolean} Determines if the provided path matches the pattern.
+ * @ignore
+ * @private
+ */
+function matchPattern(filePath, pattern, matchParent) {
+  if (pattern.paths.has(filePath)) {
+    return true;
+  }
+
+  if (matchParent) {
+    for (const childPath of pattern.paths) {
+      if (isPathInside(childPath, filePath)) {
+        return true;
       }
-    };
+    }
+  }
 
-    this.watchFilesSet.clear();
-    for (const pattern of this.watchFiles) {
-      glob.sync(pattern, globOpts).forEach(pathToCheck => {
-        if (watchIgnoreSet.has(pathToCheck)) {
-          return;
-        }
-        this.watchFilesSet.add(pathToCheck);
-      });
+  for (const globPath of pattern.globs) {
+    if (
+      minimatch(filePath, globPath, {dot: true, windowsPathsNoEscape: true})
+    ) {
+      return true;
     }
   }
 
-  has(filePath) {
-    return this.watchFilesSet.has(filePath)
+  return false;
+}
+
+/**
+ * Creates an object for matching allowed or ignored file paths.
+ * @param {PathFilter} allowed The filter for allowed paths.
+ * @param {PathFilter} ignored The filter for ignored paths.
+ * @param {string} basePath The path where mocha is run (e.g., current working directory).
+ * @returns {PathMatcher} The object for matching paths.
+ * @ignore
+ * @private
+ */
+exports.createPathMatcher = (allowed, ignored, basePath) => {
+  debug('creating path matcher from allowed: %o, ignored: %o', allowed, ignored);
+
+  /**
+   * Cache of known file paths processed by `matcher.allow()`.
+   * @type {Map<string, boolean>}
+   */
+  const allowCache = new Map();
+
+  /**
+   * Cache of known file paths processed by `matcher.ignore()`.
+   * @type {Map<string, boolean>}
+   */
+  const ignoreCache = new Map();
+
+  const MAX_CACHE_SIZE = 1000;
+
+  /**
+   * Performs a `map.set()` but will clear the map whenever the limit is reached.
+   * @param {Map<string, boolean>} map The map to use.
+   * @param {string} key The key to use.
+   * @param {boolean} value The value to set.
+   */
+  function cache(map, key, value) {
+    if (map.size >= MAX_CACHE_SIZE) {
+      map.clear();
+    }
+    map.set(key, value);
   }
+
+  /**
+   * @type {PathMatcher}
+   */
+  const matcher = {
+    allow(filePath) {
+      let allow = allowCache.get(filePath);
+      if (allow !== undefined) {
+        return allow;
+      }
+
+      allow = matchPattern(filePath, allowed.match);
```

Comment:
Why not call this and quickly check within the folders before calling `minimatch`?

```suggestion
      allow = matchPattern(filePath, allowed.match, true);
```